src/Test.Utilities/CodeFixRunner.cs

```diff
@@ -0,0 +1,299 @@
+                }
+            }
+
+            return solution;
+        }
+
+        public Solution ApplyFixesOneByOne(
+            Solution solution,
+            IEnumerable<TestAdditionalDocument> additionalFiles,
+            bool allowNewCompilerDiagnostics)
+        {
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = solution.GetProject(projectId);
+                var compilation = project.GetCompilationAsync().Result;
+                var analyzerDiagnostics = GetSortedDiagnostics(compilation, additionalFiles);
+                var compilerDiagnostics = compilation.GetDiagnostics();
+                var fixableDiagnostics = _getFixableDiagnostics(analyzerDiagnostics.Concat(compilerDiagnostics));
+
+                var diagnosticIndexToFix = 0;
+                while (diagnosticIndexToFix < fixableDiagnostics.Length)
+                {
+                    var actions = new List<CodeAction>();
+                    var diagnostic = fixableDiagnostics[diagnosticIndexToFix];
+                    var document = FindDocument(diagnostic, project);
+                    var context = new CodeFixContext(document, diagnostic, (a, d) => actions.Add(a), CancellationToken.None);
+                    _codeFixProvider.RegisterCodeFixesAsync(context).Wait();
+
+                    if (!actions.Any())
+                    {
+                        break;
+                    }
+
+                    solution = DiagnosticFixerTestsExtensions.Apply(actions.ElementAt(0));
+
+                    project = solution.GetProject(projectId);
+                    additionalFiles = project.AdditionalDocuments.Select(a => new TestAdditionalDocument(a));
+
+                    compilation = project.GetCompilationAsync().Result;
+                    analyzerDiagnostics = GetSortedDiagnostics(compilation, additionalFiles);
+
+                    var updatedCompilerDiagnostics = project.GetCompilationAsync().Result.GetDiagnostics();
+                    if (!allowNewCompilerDiagnostics)
+                    {
+                        CheckNewCompilerDiagnostics(project, compilerDiagnostics, updatedCompilerDiagnostics);
+                    }
+
+                    var newFixableDiagnostics = _getFixableDiagnostics(analyzerDiagnostics.Concat(updatedCompilerDiagnostics));
+                    if (fixableDiagnostics.SetEquals(newFixableDiagnostics, DiagnosticComparer.Instance))
+                    {
+                        diagnosticIndexToFix++;
+                    }
+                    else
+                    {
+                        fixableDiagnostics = newFixableDiagnostics;
+                        diagnosticIndexToFix = 0;
+                    }
+                }
+            }
+
+            return solution;
+        }
+
+        private static Document FindDocument(Diagnostic diagnostic, Project project)
+        {
+            foreach (var document in project.Documents)
```

Comment:
Add an assert that `diagnostic.Location.SourceTree != null`. Project level diagnostics have no source tree or location, but we don't support code fixes for them. Assert should guard against someone accidentally attempting to invoke code fix on project diagnostic.